### PR TITLE
Modernize step 30

### PR DESCRIPTION
**Pull request by drwells**
Part of #6774.

### CODE REVIEW

File: examples/step-30/doc/intro.dox [hunk 1]
```diff
@@ -483,8 +481,4 @@ extended to cover $[-1,1]\times[0,1]$ in 2D, where the flow field $\beta$ descri
 counterclockwise quarter circle around the origin in the right half of the
 domain and is parallel to the x-axis in the left part of the domain. The inflow
 boundary is again located at $x=1$ and along the positive part of the x-axis,
-and the boundary conditions are chosen as in step-12. Compared to step-12 we
-only use the more effective second assembling technique. In order to make
-comparisons more effective, we decided to keep function names like @p
-assemble_system2 even if there is only one of these routines in this tutorial
-program.
+and the boundary conditions are chosen as in step-12.
```
Comment (bangerth) [inline examples/step-30/doc/intro.dox]: nice catch!

File: examples/step-30/step-30.cc [hunk 2]
```diff
@@ -309,23 +295,19 @@ namespace Step30
...
   {
   public:
     DGMethod(const bool anisotropic);
-    ~DGMethod();
 
     void run();
 
   private:
     void setup_system();
-    void assemble_system1();
```
Comment (bangerth) [inline examples/step-30/step-30.cc]: Ah, nice. I had noticed a while ago that `assemble_system1()` is declared but never defined here. But then forgot about it again before doing anything about it. Thanks for catching it!

File: examples/step-30/step-30.cc [hunk 3]
```diff
@@ -697,12 +670,13 @@ namespace Step30
...
-                                                   endc = dof_handler.end();
-    for (unsigned int cell_no = 0; cell != endc; ++cell, ++cell_no)
-      gradient_indicator(cell_no) *=
-        std::pow(cell->diameter(), 1 + 1.0 * dim / 2);
+    unsigned int cell_n = 0;
+    for (const auto &cell : triangulation.active_cell_iterators())
+      {
+        gradient_indicator[cell_n] *=
+          std::pow(cell->diameter(), 1 + 1.0 * dim / 2);
+        ++cell_n;
```
Comment (bangerth) [inline examples/step-30/step-30.cc]: get rid of `cell_n` altogether. It is equal to `cell->active_cell_index()`, so you don't need the variable and you can all make it a one-line loop again.

File: examples/step-30/step-30.cc [hunk 4]
```diff
@@ -764,16 +734,14 @@ namespace Step30
                face_no < GeometryInfo<dim>::faces_per_cell;
                ++face_no)
             {
-              typename DoFHandler<dim>::face_iterator face =
-                cell->face(face_no);
+              auto face = cell->face(face_no);
```
Comment (bangerth) [inline examples/step-30/step-30.cc]: const

File: examples/step-30/step-30.cc [hunk 5]
```diff
@@ -764,16 +734,14 @@ namespace Step30
...
 
               if (!face->at_boundary())
                 {
                   Assert(cell->neighbor(face_no).state() ==
                            IteratorState::valid,
                          ExcInternalError());
-                  typename DoFHandler<dim>::cell_iterator neighbor =
-                    cell->neighbor(face_no);
+                  auto neighbor = cell->neighbor(face_no);
```
Comment (bangerth) [inline examples/step-30/step-30.cc]: const

File: examples/step-30/step-30.cc [hunk 6]
```diff
@@ -795,10 +763,9 @@ namespace Step30
                         {
                           // get an iterator pointing to the cell behind the
                           // present subface...
-                          typename DoFHandler<dim>::cell_iterator
-                            neighbor_child =
-                              cell->neighbor_child_on_subface(face_no,
-                                                              subface_no);
+                          auto neighbor_child =
```
Comment (bangerth) [inline examples/step-30/step-30.cc]: const

File: examples/step-30/step-30.cc [hunk 7]
```diff
@@ -918,8 +885,8 @@ namespace Step30
             }
           // Now we analyze the size of the mean jumps, which we get dividing
           // the jumps by the measure of the respective faces.
-          double average_jumps[dim];
-          double sum_of_average_jumps = 0.;
+          std::array<double, dim> average_jumps;
```
Comment (bangerth) [inline examples/step-30/step-30.cc]: not wrong, but why?

Comment (drwells) [reply examples/step-30/step-30.cc]: Its always good to advertise modern features!

**Comment by drwells** [comment]
I made the recommended changes.

**Comment by bangerth** [comment]
The tester failed because you still have this:
```
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8044/examples/step-30/step-30.cc:673:18: error: unused variable 'cell_n' [-Werror,-Wunused-variable]

    unsigned int cell_n = 0;

                 ^

```

**Comment by drwells** [comment]
whoops. Good thing we have `-Werror` to catch these silly things.

**Pull request merged**